[jit] Add None parameter as parameter instead of attributes

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33277 [WIP][quant][graphmode] Observing input/output values in callsite
* #33276 [quant][graphmode][refactor] Move check for weight outside of insertObserverFor
* #33275 [quant][graphmode][refactor] Move values_to_skip check inside valueNeedsToBeQuantized
* #33274 [quant][graphmode][refactor] Simplify signature for insertObserverFor
* #33273 [quant][graphmode][refactor] Checks for bias and weight
* #33173 [quant][graphmode] Add a pass to handle ops that doesn't require observation
* #33172 [quant][graphmode][refactor] Factor out insertDequantCall
* #33171 [quant][graphmode] refactor nodeQuantizable
* #32816 [quant][graphmode] Add add_relu pattern in skip values
* #32815 [quant][graphmode] Add asserts for Conv and ReLU modules
* #32814 [quant][graphmode] Skip quantizing input and output in matched module
* #32813 [quant][graphmode][refactor] Separate preprocess step for insertObserver
* #32812 [quant][graphmode][refactor] Change signature of getModuleAccessPath
* #32809 [quant][graphmode][refactor] Move the check for qconfig inside insertObserver call
* #32379 [quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes
* **#32964 [jit] Add None parameter as parameter instead of attributes**
* #32846 [jit][fix] Remove slot in parameter slot

Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D19913188](https://our.internmc.facebook.com/intern/diff/D19913188)

jerryzh168 committed:
[jit] Add None parameter as parameter instead of attributes

Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

suo reviewed (changes requested):
Can you add some tests exercising the new `parameters()` `script::Module` API when there are some none parameters?

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -367,7 +367,7 @@ struct slot_iterator_impl {
   bool valid() const {
     return top().i_ < int64_t(top().module_._ivalue()->type()->numAttributes()) &&
-        Policy::valid(top().module_._ivalue()->type(), top().i_);
+      Policy::valid(top().module_._ivalue()->type(), top().i_, top().module_._ivalue()->getSlot(top().i_));
```
nit: please clang-format your changes :)

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -478,8 +478,8 @@ struct TORCH_API ParameterPolicy {
-  static bool valid(const ClassTypePtr& typ, size_t i) {
-    return typ->is_parameter(i);
+  static bool valid(const ClassTypePtr& typ, size_t i, const IValue& v) {
+    return typ->is_parameter(i) && v.isTensor();
```
I think it's also worth adding some enforcement to `ClassType`: if we're trying to add a parameter, it must be Tensor, Optional[Tensor], or NoneType.

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

suo commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -70,6 +71,14 @@ size_t ClassType::addAttribute(
+             type->expect<OptionalType>()->getElementType()->kind() ==
+                 TensorType::Kind) ||
+            (type->kind() == NoneType::Kind),
+        "Expecting parameter to have either None, Tensor or Optional[Tensor] type, but got:",
```
nit: space after `got:`

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D19913188](https://our.internmc.facebook.com/intern/diff/D19913188)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D19913188](https://our.internmc.facebook.com/intern/diff/D19913188)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add None parameter as parameter instead of attributes"


Summary:
att

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D19913188](https://our.internmc.facebook.com/intern/diff/D19913188)

[ghstack-poisoned]
